Sources/performance-worker/WorkerService.swift

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import GRPCCore
+import NIOConcurrencyHelpers
+import NIOCore
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+final class WorkerService: Grpc_Testing_WorkerService.ServiceProtocol, Sendable {
+  let state: NIOLockedValueBox<State>
+
+  init() {
+    let clientAndServer = State()
+    self.state = NIOLockedValueBox(clientAndServer)
+  }
+
+  struct State {
+    var role: Role?
+
+    enum Role {
+      case client(GRPCClient)
+      case server(GRPCServer)
+    }
+
+    init() {}
+
+    init(role: Role) {
+      self.role = role
+    }
+
+    init(server: GRPCServer) {
+      self.role = .server(server)
+    }
+
+    init(client: GRPCClient) {
+      self.role = .client(client)
+    }
+  }
+
+  func quitWorker(
+    request: ServerRequest.Single<Grpc_Testing_WorkerService.Method.QuitWorker.Input>
+  ) async throws -> ServerResponse.Single<Grpc_Testing_WorkerService.Method.QuitWorker.Output> {
+
+    if let role = self.state.withLockedValue({ $0.role }) {
```

Comment:
I don't think there's any reason to delay setting the `$0.role` back to `nil`, we can just do it in one go:

```swift
let role = self.state.withLockedValue { 
  defer { $0.role = nil }
  return $0.role
}
```